bytecomp/matching.ml

```diff
@@ -2881,29 +2884,66 @@ and do_compile_matching repr partial ctx arg pmh = match pmh with
         (divide_record lbl.lbl_all (normalize_pat pat))
         ctx_combine repr partial ctx pm
   | Tpat_constant cst ->
+      let names = None in
       compile_test
         (compile_match repr partial) partial
         divide_constant
-        (combine_constant pat.pat_loc arg cst partial)
+        (combine_constant names pat.pat_loc arg cst partial)
         ctx pm
-  | Tpat_construct (_, cstr, _) ->
+  | Tpat_construct (_, cstr, pats) ->
+      let names_from_type_variant cstrs =
+        let (consts, blocks) = List.fold_left
+          (fun (consts, blocks) cstr ->
+            if cstr.Types.cd_args = []
+            then (Ident.name cstr.Types.cd_id :: consts, blocks)
+            else (consts, Ident.name cstr.Types.cd_id :: blocks))
+          ([], []) cstrs in
+        Some {consts = consts |> List.rev |> Array.of_list;
+              blocks = blocks |> List.rev |> Array.of_list } in
+
+      let rec resolve_path n path =
+        match Env.find_type path pat.pat_env with
+        | {type_kind = Type_variant cstrs} ->
+          names_from_type_variant cstrs
+        | {type_kind = Type_abstract; type_manifest = Some t} ->
+          ( match (Ctype.unalias t).desc with
+            | Tconstr (pathn, _, _) ->
+              (* Format.eprintf "XXX path%d:%s path%d:%s@." n (Path.name path) (n+1) (Path.name pathn); *)
+              resolve_path (n+1) pathn
+            | _ -> assert false)
+        | {type_kind = Type_abstract; type_manifest = None} ->
+          (* Format.eprintf "XXX Type_abstract@."; *)
+          Some {consts=[||]; blocks=[||]}
+        | {type_kind = Type_record _} ->
+          (* Format.eprintf "XXX Type_record@."; *)
+          Some {consts=[||]; blocks=[||]}
+        | {type_kind = Type_open } ->
+          (* Format.eprintf "XXX Type_open@."; *)
+          Some {consts=[||]; blocks=[||]} in
+
+      let names = match (Btype.repr pat.pat_type).desc with
+        | Tconstr (path, _, _) -> resolve_path 0 path
+        | _ -> assert false in
```

Comment:
would `assert false` be too harsh?

Reply:
I've tried to be intentionally harsh here to make sure we have everything covered, during testing.

Reply:
cleaned up now